### PR TITLE
[server] Fix domain scope for session cookie

### LINKED ISSUE #1887: Self hosted gitpod and gitlab Auth issues
Hi

i've installed gitpod (non-ssl) onto a local box,     and tried to integrate into our local gitlab instance (non-ssl)
when I hit the login button it redirects to gitlab,  accessing for authorisation and then the callback to gitpod occurs,   however gitpod just says 

"Sorry, something went wrong 😓
Please allow Cookies in your browser and try to log in again."

disabled all extension and allowed cookies,  

Server logs show

User started the login process
(GitLab Auth) No session found during auth callback

### PR DESCRIPTION
**Pull request by AlexTugarev**
With this change slicing of hostname will only be applied for preview environments.

ACK this still only works for preview environments deployed without a 2nd level TLD.
OTOH it should quickly enable SH installations with 2nd level TLDs!

Fixes #1887

### CODE REVIEW

**Comment by jankeromnes** [comment]
Failed with:

```
  1) UserDBSpec
       "before each" hook for "createUserAndFindById":
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

We should really figure out how to configure the timeout for the "before each" hook in our tests, because at 2s they time out quite often.

/werft run

**Comment by jankeromnes** [comment]
I tried this, but not sure if relevant:

1. Log in via GitHub
2. Go to Access Control
3. Click on "Connect" for GitLab

Result:

Screenshot 2020-09-25 at 16 51 11

**Comment by jankeromnes** [comment]
Update: Doing the same thing in Chrome works:

Screenshot 2020-09-25 at 16 53 52

**Comment by jankeromnes** [comment]
Maybe this is because we use insecure HTTP deployments while we should be using HTTPS instead?

**Comment by AlexTugarev** [comment]
> Note: Using Firefox Nightly as my browser.


@jankeromnes does it also fail with regular FF?

**Comment by hm2075** [comment]
build failing?

**Comment by hm2075** [comment]
can i request for this to be bumped up?

**Comment by hm2075** [comment]
tested this

image

**Comment by AlexTugarev** [comment]
@hm2075 thanks for watching this issue!

what exactly did you test? did you deploy the server image from this branch?

> would that happen because my instance is behind a corporate firewall?

I have no clue about the setup of your FW.  

The OAuth error OTOH looks unrelated to the underlying issue, though. This indeed indicates a broken config, most likely on the auth server side, e.g. wrong return URL or client id/password.

**Comment by hm2075** [comment]
> @hm2075 thanks for watching this issue!
> 
> what exactly did you test? did you deploy the server image from this branch?
> 
> > would that happen because my instance is behind a corporate firewall?
> 
> I have no clue about the setup of your FW.
> 
> The OAuth error OTOH looks unrelated to the underlying issue, though. This indeed indicates a broken config, most likely on the auth server side, e.g. wrong return URL or client id/password.

hi

i just added my local gitlab lab server to  http://at-session-fix.staging.gitpod-dev.com/

probably not the best way to test this,   however testing with gitlab.com worked fine


I guess i would need to deploy this fix to my local instance to test fully

File: components/server/src/session-handler.ts [hunk 1]
```diff
@@ -43,11 +43,25 @@ export class SessionHandlerProvider {
...
+            //
+            // Otherwise, clients (in this case Chrome) may ignore (as in: save it, but don't send it on consequent requests) the 'Set-Cookie:...' send with a redirect (302, to github oauth)
+            // For details, see:
+            // - RFC draft sameSite: http://httpwg.org/http-extensions/draft-ietf-httpbis-cookie-same-site.html
+            // - https://bugs.chromium.org/p/chromium/issues/detail?id=150066
+            // - google: chromium not sending cookies set with redirect
+            
+            const hostParts = hostName.split('.');
+            const baseDomain = hostParts.slice(hostParts.length - 2).join('.');
+            domain = `.${baseDomain}`;
```
Comment (svenefftinge) [inline components/server/src/session-handler.ts:60]: I think we need the dot prefixing in all cases, as we want tosend the cookie with requests to subdomains as well, no?

File: components/server/src/session-handler.ts [hunk 2]
```diff
@@ -43,11 +43,25 @@ export class SessionHandlerProvider {
...
+            // Use cookie for base domain to allow cookies being sent via ingress proxy in preview environments
+            //
+            // Otherwise, clients (in this case Chrome) may ignore (as in: save it, but don't send it on consequent requests) the 'Set-Cookie:...' send with a redirect (302, to github oauth)
+            // For details, see:
+            // - RFC draft sameSite: http://httpwg.org/http-extensions/draft-ietf-httpbis-cookie-same-site.html
+            // - https://bugs.chromium.org/p/chromium/issues/detail?id=150066
+            // - google: chromium not sending cookies set with redirect
+            
+            const hostParts = hostName.split('.');
+            const baseDomain = hostParts.slice(hostParts.length - 2).join('.');
```
Comment (svenefftinge) [inline components/server/src/session-handler.ts:59]: I wonder why the slicing and special handling is necessary here. Also in dev staging all the requests should go to the hostname URL. Is that not the case?

File: components/server/src/session-handler.ts (hunk 1 above)
Comment (AlexTugarev) [reply components/server/src/session-handler.ts:60]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie

cf. Domain

> Contrary to earlier specifications, leading dots in domain names (.example.com) are ignored.

Comment (svenefftinge) [reply components/server/src/session-handler.ts:60]: then lets remove the dot here as well.

**Review by svenefftinge — APPROVED** [review]

**Comment by hm2075** [comment]
@AlexTugarev 

making good progress,  I managed to get a copy of session fix 7 and run it on our kubernetes install,      it looks like the revised code now allows me to kind of authenticate,  before I was getting cookie errors,  

i setup two auth providers,   gitlab.com  and our selfhosted instance


gitlab.com works fine

for some reason our local one is coming back with an Oauth Error

Failed to obtain access token\n    at GenericOAuth2Strategy.OAuth2Strategy._createOAuthError (/app/node_modules/passport-oauth2/lib/strategy.js:379:17)\n    at /app/node_modules/passport-oauth2/lib/strategy.js:166:45\n    at patchedCallback (/app/node_modules/@gitpod/server/dist/src/auth/generic-auth-provider.js:939:28)\n    at /app/node_modules/oauth/lib/oauth2.js:191:18\n    at ClientRequest.<anonymous> (/app/node_modules/oauth/lib/oauth2.js:162:5)\n    at ClientRequest.emit (events.js:315:20)\n    at ClientRequest.EventEmitter.emit (domain.js:483:12)\n    at TLSSocket.socketErrorListener (_http_client.js:426:9)\n    at TLSSocket.emit (events.js:315:20)\n    at TLSSocket.EventEmitter.emit (domain.js:483:12)\n    at emitErrorNT (internal/streams/destroy.js:92:8)\n    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)\n    at processTicksAndRejections (internal/process/task_queues.js:84:21)","payload":{"authorize-flow":false,"login-flow":true,"clientInfo":{"ua":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36","fingerprint":"f34c9db3168b457c6edbff0ac6965b37c9c0d5a932b243a431e358075e3b8813"},"authProviderId":"Nexus-GitLab","err":"



any pointers?

File: components/server/src/session-handler.ts (hunk 2 above)
Comment (AlexTugarev) [reply components/server/src/session-handler.ts:59]: exactly. I recall, why this is actually needed. in dev staging preview environments we have the ingress proxy which terminates https connections, which is a crucial part in the login process, right? the URL of the ingress proxy has no namespace prefix at all, that's why we've chosen the sliced version of it. this allows those cookies to be present on  ingress proxy requests, which are then forwarded to the services.

**Pull request merged**